Rwlock sanitary update 1: removed weak_ptrs from implementation + added strong exception safety guarantee

Akazz opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Improvement

Short description (up to few sentences):
 - The most important change is removing weak_ptrs from implementation. As it is the primary source of glitches during destruction of `RWLockImpl::LockHolderImpl` objects.
 - Strong exception safety is a very important property for this key primitive: for our data to be safe RWLock's state must be preserved consistent at all times, especially during high memory pressure when `std::bad_alloc` is a common system's response

Commit:
Changed FastPath + better diagnostics

Commit:
Cosmetics

Commit:
Better

Commit:
Refactored Clients -> Referers

Commit:
WIP - About to give up storing weak_ptrs!

Commit:
Dropped weak_ptrs!

Commit:
Removed some garbage

Commit:
Minor changes

Commit:
Cosmetic stuff

alexey-milovidov commented on dbms/src/Common/RWLock.cpp:
```diff
@@ -201,33 +203,33 @@ RWLockImpl::LockHolderImpl::~LockHolderImpl()
 
-    /// Remove the group if we were the last client and notify the next group
-    if (it_group->clients.empty())
+    /// Remove the group if we are the last referer and notify the next group
```
Contrary to the HTTP standard, the `referrer` word is written with two `r`.

Akazz replied:
Good point. Corrected this

Commit:
Strong exception safety guarantee:
Based on std::list<>::emplace_back() and std::unordered_map<>::emplace()
provide strong exception safety, RWLockImpl is now changed to provide
the same level of exception safety.

Commit:
Dropped some garbage lines

Commit:
Some formatting + minor change

Commit:
Build fixed
